### PR TITLE
[Tooltip] - add component

### PR DESCRIPTION
**Pull request by pietro-maximoff**
https://sovryn.atlassian.net/browse/SOV-146

### CODE REVIEW

File: packages/ui/src/2_molecules/Tooltip/Tooltip.tsx [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  ]);
+
+  useEffect(() => {
+    if (!isVisible) {
+      return;
+    }
+    const coords = getCoords();
+    setCoords(coords);
+    onShow?.();
+  }, [isVisible, getCoords, placement, setCoords]);
```
Comment (creed-victor) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.tsx]: I somehow think that this should be part of the `onMouseEnter`, or it causes some kind of problems?

Comment (pietro-maximoff) [reply packages/ui/src/2_molecules/Tooltip/Tooltip.tsx]: this effect updates elements data (tooltip and target) when you trigger the tooltip by [click, hover, focus]

**Comment by tiltom** [comment]
@pietro-maximoff I was testing it at the PR env and positioning is not working correctly, see the screenshot.
image

**Comment by Rickk137** [comment]
@pietro-maximoff the TooltipTrigger.click when tooltip is opened it only closes on click to the element. I think clicking outside should also close the tooltip.

**Comment by pietro-maximoff** [comment]
> @pietro-maximoff the TooltipTrigger.click when tooltip is opened it only closes on click to the element. I think clicking outside should also close the tooltip.

Yep, I think it's a good point, will add it

**Comment by pietro-maximoff** [comment]
> @pietro-maximoff the TooltipTrigger.click when tooltip is opened it only closes on click to the element. I think clicking outside should also close the tooltip.

Yep, I think it's a good point, will add it

File: packages/ui/src/2_molecules/Tooltip/Tooltip.module.css [hunk 2]
```diff
@@ -0,0 +1,88 @@
+.tooltip {
+    @apply bg-gray-10 rounded p-3 text-gray-1 absolute inline-block min-w-5 max-w-72 leading-5 opacity-0;
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.module.css]:
```suggestion
    @apply bg-gray-10 rounded p-3 text-gray-90 absolute inline-block min-w-5 max-w-72 leading-5 opacity-0;
```
Tooltip text should be `gray-90` according to the design.

File: packages/ui/src/2_molecules/Tooltip/Tooltip.module.css [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        &-end {
+            &:after {
+                @apply left-4 right-auto;
+            }
+        }
+        &-start {
+
+            &:after {
+                @apply left-auto right-4;
+            }
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.module.css]: I think you have `-end` and `-start` variation the other way around (so in this case you should position `-start` variation from `left` and not `right`). It's not critical, just looks a bit weird in the code :D

File: packages/ui/src/2_molecules/Tooltip/Tooltip.test.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+  afterEach(() => {
+    jest.useRealTimers();
+  });
+
+  test('should render a tooltip on hover in and hide on hover out', () => {
+    render(<Tooltip children={<button>Text</button>} content={<>Tooltip</>} />);
+    const button = screen.getByRole('button');
+    userEvent.hover(button);
+    const tooltip = screen.queryByText('Tooltip');
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.test.tsx]:
```suggestion
    const { getByRole, queryByText } = render(<Tooltip children={<button>Text</button>} content={<>Tooltip</>} />);
    const button = getByRole('button');
    userEvent.hover(button);
    const tooltip = queryByText('Tooltip');
```

File: packages/ui/src/2_molecules/Tooltip/Tooltip.test.tsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    render(
+      <Tooltip
+        trigger={TooltipTrigger.focus}
+        children={<button>Text</button>}
+        content={<>Tooltip</>}
+      />,
+    );
+    const button = screen.getByRole('button');
+    await waitFor(() => button.focus());
+    const tooltip = screen.queryByText('Tooltip');
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.test.tsx]:
```suggestion
    const { getByRole, queryByText } = render(
      <Tooltip
        trigger={TooltipTrigger.focus}
        children={<button>Text</button>}
        content={<>Tooltip</>}
      />,
    );
    const button = getByRole('button');
    await waitFor(() => button.focus());
    const tooltip = queryByText('Tooltip');
```

File: packages/ui/src/2_molecules/Tooltip/Tooltip.test.tsx [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    render(
+      <Tooltip
+        trigger={TooltipTrigger.click}
+        children={<button>Text</button>}
+        content={<>Tooltip</>}
+      />,
+    );
+    const button = screen.getByRole('button');
+    userEvent.click(button);
+    const tooltip = screen.queryByText('Tooltip');
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.test.tsx]:
```suggestion
      const { getByRole, queryByText } = render(
      <Tooltip
        trigger={TooltipTrigger.click}
        children={<button>Text</button>}
        content={<>Tooltip</>}
      />,
    );
    const button = getByRole('button');
    userEvent.click(button);
    const tooltip = queryByText('Tooltip');
```

File: packages/ui/src/2_molecules/Tooltip/Tooltip.test.tsx [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  it('should not render a tooltip if disabled is true', () => {
+    render(
+      <Tooltip
+        disabled
+        children={<button>Text</button>}
+        content={<>Tooltip</>}
+      />,
+    );
+    userEvent.click(screen.getByRole('button'));
+    expect(screen.queryByText('Tooltip')).not.toBeInTheDocument();
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.test.tsx]:
```suggestion
      const { getByRole, queryByText } = render(
      <Tooltip
        disabled
        children={<button>Text</button>}
        content={<>Tooltip</>}
      />,
    );
    userEvent.click(getByRole('button'));
    expect(queryByText('Tooltip')).not.toBeInTheDocument();
```

File: packages/ui/src/2_molecules/Tooltip/Tooltip.test.tsx [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  it('should render a tooltip with a bottom placement', () => {
+    render(
+      <Tooltip
+        children={<button>Text</button>}
+        placement={TooltipPlacement.BOTTOM}
+        content={<>Tooltip</>}
+      />,
+    );
+    userEvent.hover(screen.getByRole('button'));
+    const tooltip = screen.getByText('Tooltip');
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.test.tsx]:
```suggestion
      const { getByRole, queryByText } = render(
      <Tooltip
        children={<button>Text</button>}
        placement={TooltipPlacement.BOTTOM}
        content={<>Tooltip</>}
      />,
    );
    userEvent.hover(getByRole('button'));
    const tooltip = getByText('Tooltip');
```

File: packages/ui/src/2_molecules/Tooltip/Tooltip.types.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
+export enum TooltipPlacement {
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.types.ts:1]: It seems like we haven't been 100% consistent with this but I'd suggest to use `camelCase` naming in `enums` (e.g. `bottomEnd, bottomStart, bottom, ...`). I was looking at our [Code Guidelines](https://sovryn.monday.com/docs/2733202752) and it says we should use `UpperCamelCase` but I believe that's for just `enum` names and not values.

 What do you think @soulBit ?

**Review by creed-victor — APPROVED** [review]

File: packages/ui/src/2_molecules/Tooltip/Tooltip.tsx [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+  trigger = TooltipTrigger.hover,
+}) => {
+  const tooltipRef = useRef<HTMLDivElement>(null);
+  const targetRef = useRef<HTMLElement>(null);
+  const [elements, setElements] = useState<Nullable<TooltipElements>>(null);
+  const [isVisible, setIsVisible] = useState(false);
+  const [isHovered, setIsHovered] = useState(false);
+  const [shouldHide, setShouldHide] = useState(false);
+  const isHoveredRef = useRef(isHovered);
+  isHoveredRef.current = isHovered;
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.tsx]:
```suggestion

```

File: packages/ui/src/2_molecules/Tooltip/Tooltip.tsx [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    if (target && tooltip) {
+      return { target, tooltip };
+    }
+    return null;
+  }, []);
+
+  const updateElements = useCallback(() => {
+    const elements = getElements();
+    setElements(elements);
+  }, [getElements, setElements]);
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.tsx]:
```suggestion
  const updateElements = useCallback(
    () => setElements(getElements()),
    [getElements, setElements],
  );
```

File: packages/ui/src/2_molecules/Tooltip/Tooltip.tsx [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+  const handleShow = useCallback(() => {
+    setIsVisible(prevValue => !prevValue);
+    if (isVisible) {
+      setShouldHide(true);
+    }
+  }, [setIsVisible, setShouldHide, isVisible]);
+
+  const handleHide = useCallback(() => {
+    setShouldHide(true);
+  }, [setShouldHide]);
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.tsx]:
```suggestion
  const handleHide = useCallback(() => setShouldHide(true), [setShouldHide]);
```

File: packages/ui/src/2_molecules/Tooltip/Tooltip.tsx [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+
+  const onMouseHover = useCallback(() => {
+    setIsHovered(prevValue => !prevValue);
+  }, [setIsHovered]);
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.tsx]:
```suggestion
  const onMouseHover = useCallback(
    () => setIsHovered(prevValue => !prevValue),
    [setIsHovered],
  );
```

File: packages/ui/src/2_molecules/Tooltip/Tooltip.tsx [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+
+  const getElementProps = useCallback(() => {
+    const attributes = {
+      'data-layout-id': dataLayoutId,
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.tsx]: I'd create `DATA_ATTRIBUTE` constant in `types/index.tsx` so we don't need to hardcode `data-layout-id` every time we need to use it as string.

Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.tsx]:
```suggestion
      [DATA_ATTRIBUTE]: dataLayoutId,
```

**Review by tiltom — CHANGES_REQUESTED** [review]

File: packages/ui/src/2_molecules/Tooltip/Tooltip.module.css (hunk 2 above)
Comment (pietro-maximoff) [reply packages/ui/src/2_molecules/Tooltip/Tooltip.module.css]: yep, seems it was updated later

File: packages/ui/src/2_molecules/Tooltip/Tooltip.types.ts (hunk 9 above)
Comment (soulBit) [reply packages/ui/src/2_molecules/Tooltip/Tooltip.types.ts:1]: @pietro-maximoff let's go with camelCase enum keys as suggested, please can you update this enum and the code guidelines doc to match?

File: packages/ui/src/2_molecules/Tooltip/Tooltip.tsx (hunk 14 above)
Comment (soulBit) [reply packages/ui/src/2_molecules/Tooltip/Tooltip.tsx]: @tiltom good suggestion, but let's also apply it to the other components after this PR is merged - added new story for this here: https://sovryn.atlassian.net/browse/SOV-861

File: packages/ui/src/2_molecules/Tooltip/Tooltip.types.ts (hunk 9 above)
Comment (tiltom) [reply packages/ui/src/2_molecules/Tooltip/Tooltip.types.ts:1]: @pietro-maximoff sorry I wasn't specific enough in my comment but as @soulBit mentioned, this is mainly about `keys` (`values` can stay as they are now) so change it pls (e.g. `BOTTOM_END` will be `bottomEnd`).

File: packages/ui/src/2_molecules/Tooltip/Tooltip.tsx [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+    () => setElements(getElements()),
+    [getElements, setElements],
+  );
+
+  const tooltipPosition = useMemo(() => {
+    if (!elements) {
+      return;
+    }
+    return getTooltipPosition(elements, placement);
+  }, [elements, placement]);
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.tsx]:
```suggestion
  const tooltipPosition = useMemo(() => {
    if (elements) {
      return getTooltipPosition(elements, placement);
    }
  }, [elements, placement]);
```

File: packages/ui/src/2_molecules/Tooltip/Tooltip.tsx [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+  }, [setIsVisible, setShouldHide, isVisible]);
+
+  const handleHide = useCallback(() => setShouldHide(true), [setShouldHide]);
+
+  const onMouseHover = useCallback(
+    () => setIsHovered(prevValue => !prevValue),
+    [setIsHovered],
+  );
+
+  const getElementProps = useCallback(() => {
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.tsx]:
```suggestion
  const elementProps = useMemo(() => {
```
This is just my personal preference but I think it's semantically more correct to use `useMemo` here as you just return a memoized object from this method and not an actuall callback.

File: packages/ui/src/2_molecules/Tooltip/Tooltip.tsx [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+    };
+  }, [updateElements]);
+
+  useEffect(() => {
+    if (!isVisible) {
+      return;
+    }
+    updateElements();
+    onShow?.();
+  }, [isVisible, updateElements, onShow]);
```
Comment (tiltom) [inline packages/ui/src/2_molecules/Tooltip/Tooltip.tsx]:
```suggestion
  useEffect(() => {
    if (isVisible) {
      updateElements();
      onShow?.();
    }
  }, [isVisible, updateElements, onShow]);
```

**Review by tiltom — CHANGES_REQUESTED** [review]

**Review by tiltom — APPROVED** [review]

**Comment by Rickk137** [comment]
@pietro-maximoff I get this issue if the tooltip is near screen borders
Screen Shot 2022-10-18 at 2 40 17 PM

**Review by LunaSov — APPROVED** [review]

**Pull request merged**